### PR TITLE
28 ensure slug uniqueness

### PR DESCRIPTION
**Pull request by lojzatran**
Fixes: https://github.com/commercetools/commercetools-node-variant-reassignment/issues/28

### CODE REVIEW

File: spec/unit/runner/variant-reassignment.spec.js [hunk 1]
```diff
@@ -307,4 +309,62 @@ describe('Variant reassignment', () => {
...
   })
+
+  describe('ensure slug uniqueness', () => {
+    const testProductId = 'test-product-id'
+    let productsToAnonymize
+    let variantReassignments
+    beforeEach(() => {
+      productsToAnonymize = []
+      const productService = new ProductManager(utils.logger, {})
+      sinon.stub(productService, 'anonymizeCtpProduct').callsFake((product) => {
```
Comment (junajan) [inline spec/unit/runner/variant-reassignment.spec.js]: Here you can use a simple `stub.returns(Promise.resolve())` and later in the code you can check `stub. callCount` and `stub.args` so you will get rid of redundant variable `productsToAnonymize `

File: lib/runner/variant-reassignment.js [hunk 2]
```diff
@@ -394,6 +429,7 @@ export default class VariantReassignment {
     return Promise.map(Array.from(productToSkusToRemoveMap), ([product, skus]) =>
         this.productService.removeVariantsFromProduct(product, skus),
       { concurrency: 3 })
+      .then(result => _.compact(result))
```
Comment (junajan) [inline lib/runner/variant-reassignment.js]: Simplified: `.then(_.compact)`

File: lib/runner/variant-reassignment.js [hunk 3]
```diff
@@ -332,10 +333,44 @@ export default class VariantReassignment {
 
   /**
    * Verify that there are no other products in the platform that has slugs of productDraft
-   * except ctpProductToUpdate.
-   * This method should cover the test variant-reassignment-example-7.spec.js
+   * except ctpProductToUpdate. It's enough to check matchingProducts because it's not probable
+   * that there will be a product which has no matching variant, but a conflicting slug.
```
Comment (junajan) [inline lib/runner/variant-reassignment.js]: `it's not probable` .. hopefully 🙏

File: lib/runner/variant-reassignment.js [hunk 4]
```diff
@@ -332,10 +333,44 @@ export default class VariantReassignment {
...
+   * @see variant-reassignment-example-7.spec.js
+   */
+  async _ensureSlugUniqueness (productDraft, matchingProducts) {
+    const productDraftSlug = productDraft.slug
+    const productsToAnonymize = []
+    matchingProducts.forEach((product) => {
+      const isSlugConflicting = this._isSlugConflicting(product, productDraftSlug)
+      if (isSlugConflicting)
+        productsToAnonymize.push(product)
+    })
```
Comment (junajan) [inline lib/runner/variant-reassignment.js]: Simplified version:
```js
const productsToAnonymize = matchingProducts.map(product =>
  this._isSlugConflicting(product, productDraftSlug)
)
```

File: lib/runner/variant-reassignment.js [hunk 5]
```diff
@@ -332,10 +333,44 @@ export default class VariantReassignment {
...
+    const stagedSlug = product.masterData.staged.slug
+    const currentSlug = product.masterData.current.slug
+    const stagedDraftSlugs = _.merge({}, productDraftSlug, stagedSlug)
+    const currentDraftSlugs = _.merge({}, productDraftSlug, currentSlug)
+    const productDraftSlugLength = Object.keys(productDraftSlug).length
+    const stagedSlugLength = Object.keys(stagedSlug).length
+    const currentSlugLength = Object.keys(currentSlug).length
+
+    return stagedSlugLength + productDraftSlugLength > Object.keys(stagedDraftSlugs).length
+      || currentSlugLength + productDraftSlugLength > Object.keys(currentDraftSlugs).length
```
Comment (junajan) [inline lib/runner/variant-reassignment.js]: I tried to clear this method using loop:
```js
_isSlugConflicting (product, productDraftSlug) {
  const productDraftSlugLength = Object.keys(productDraftSlug).length
  
  // if at least one version has conflict in slugs, return true
  for(let version of ['staged', 'current']) {
    const slug = product.masterData[version].slug
    const slugLength = Object.keys(stagedSlug).length
    const stagedDraftSlugs = _.merge({}, productDraftSlug, slug)

    if(slugLength + productDraftSlugLength !== Object.keys(stagedDraftSlugs).length)
      return true
  }
  return false
}
```

File: lib/runner/variant-reassignment.js (hunk 3 above)
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: 🙏

File: lib/runner/variant-reassignment.js (hunk 4 above)
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: Do you mean this?
```js
const productsToAnonymize = matchingProducts.filter(product =>
      this._isSlugConflicting(product, productDraftSlug)
    )
```

File: spec/unit/runner/variant-reassignment.spec.js (hunk 1 above)
Comment (lojzatran) [reply spec/unit/runner/variant-reassignment.spec.js]: But isn't callCount only for spy? And if you have spy, then it calls the real method and you cannot fake the method?

File: lib/runner/variant-reassignment.js (hunk 4 above)
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: https://github.com/commercetools/commercetools-node-variant-reassignment/pull/37/commits/418eb9261e545b390510ddcaf12a97939aa61d93

File: lib/runner/variant-reassignment.js (hunk 5 above)
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: https://github.com/commercetools/commercetools-node-variant-reassignment/pull/37/commits/626d763a9fae27de8b2035fd1857ca6a8346856b

File: lib/runner/variant-reassignment.js (hunk 2 above)
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: https://github.com/commercetools/commercetools-node-variant-reassignment/pull/37/commits/626d763a9fae27de8b2035fd1857ca6a8346856b#diff-eda63119a30637151d12307cefd81cf0R432

File: spec/unit/runner/variant-reassignment.spec.js (hunk 1 above)
Comment (junajan) [reply spec/unit/runner/variant-reassignment.spec.js]: Try to print out stub object - I think that it extends spy functionality

Comment (lojzatran) [reply spec/unit/runner/variant-reassignment.spec.js]: https://github.com/commercetools/commercetools-node-variant-reassignment/pull/37/commits/a0772affb546dc46fcda78c5d61e0c5eed5a615d

**Comment by junajan** [comment]
Resolve conflict and then you can merge

File: lib/runner/variant-reassignment.js (hunk 3 above)
Comment (lojzatran) [reply lib/runner/variant-reassignment.js]: @junajan this assumption is now gone with the new example 16 :D

Comment (junajan) [reply lib/runner/variant-reassignment.js]: :D it didn't last long

**Pull request merged**